### PR TITLE
refactor: Improve CRollingBloomFilter::reset by using std::fill

### PR DESCRIPTION
**Pull request by promag**
Cleaner code. Also improves performance with `--enable-debug` (which is meaningless to non-developers).

Before:
```
# Benchmark, evals, iterations, total, min, max, median
RollingBloomReset, 5, 150, 19.3008, 0.0254917, 0.0259195, 0.0257395
```
After:
```
# Benchmark, evals, iterations, total, min, max, median
RollingBloomReset, 5, 150, 5.43269, 0.00720651, 0.00729697, 0.00724854
```

### CODE REVIEW

**Comment by promag** [comment]
I was going to use `memset` but I think `std::fill` is much nicer and probably as fast as `memset`.

**Comment by gmaxwell** [comment]
Seems okay to me, but why are we resetting a big rolling bloom filter this often?

It may be better to use a generation counter. (this isn't a reason not to take this improvement, but maybe an area for bigger improvement).

**Comment by MarcoFalke** [comment]
Any intuition why this is faster? Could add a trivial benchmark as first commit?

**Comment by promag** [comment]
@MarcoFalke one reason may be a specialized version of fill for uint64_t. Will add the benchmark.

**Comment by jonasschnelli** [comment]
Well spotted!
Concept ACK

**Comment by promag** [comment]
Updated OP with benchmark result.

Profiling before and after this change:

Screenshot 2019-05-22 at 00 11 14

Screenshot 2019-05-22 at 00 10 12

**Comment by promag** [comment]
@MarcoFalke benchmark added.

**Comment by laanwj** [comment]
> Any intuition why this is faster? Could add a trivial benchmark as first commit?

IIRC `std::fill` tends to be optimized to a memset, which uses arch-specific instructions for the fastest filling of memory, whereas a loop is not

**Comment by promag** [comment]
> but maybe an area for bigger improvement

@gmaxwell maybe open an issue and throw some suggestions?

**Review by jamesob — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/16073/commits/a47b5607f63f486f8576af4a89cb60cf9472caee

Nice find.

**Comment by MarcoFalke** [comment]
What compiler are you using?


----
utACK 8d774bbbe8454119161446225451210c23fe56d3

Couldn't find a performance difference on my system, but the code looks cleaner.

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

utACK 8d774bbbe8454119161446225451210c23fe56d3

Couldn't find a performance difference on my system, but the code looks cleaner.
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiX4gwAsi8srAjhAkRx0STvjJ75oiaU5ADUonn7zrstBc17F2kNONLedb+vR9J/
elm5lUmqrucGVItm1UlFgaOB/yq100B688xovcaOez2RKuSTT51y26Aur1Amrtyk
CDiP1dYabzImeSUPbaKEP33Hx0fNbzp77NAqbUW/ufJUYqAlmbh5QJsKP/baMKyo
f+1Jt5n/fDA9B9YNIWckb5qLAqlLDQE07RrMLRROcIHW3+ipsxdHEpug49QEzlTE
15lQAjzTXOfVzAX78dxdFc0lRzVr1MmPyAWlA26xhqe+4Lo4ObW7OMeSWHhMlISL
fntaqw2K8iYz8rBo218A+nY4BjcJSMEcjOAP2n1p+kU2rSm/NbyIjwZaHOv4xIWT
S/Y5DbZIZdPmIcEM/CXsGUzGENjIUeCyCa77+Rt6O0swC2OKaW3cJ02u+Iksq9iY
UnVv029AZdjRz+RGDHk2ubosddk/hczXTUju9+lUDoRpH0FKu0T5efn7Co1yT3gR
gwypjCZu
=k0jH
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `0afe9d4d13b7b6fae0b5c6625b924bdb707bd6bf085d76618e964707100296a0  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401080afe9d4d13b7b6fae0b5c6625b924bdb707bd6bf085d76618e964707100296a0f01070d4799efc4b68259c7a7b637688551908fff0104326f0dd8847a5bfb5a9385f594a41fb08f1045ce558aff0082bf13c4108de7a230083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010494157ec028be229e9196bb47138205e08f1045ce558aff0087222d3efaf9cb4750083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b5afa06859739cf362a81d6ca81d29ea08f01087e14204f94f3255b2d337abe775605c08f1045ce558aff0086903c685c8ce15b00083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010bdf02479b393453188036d495e53eae108f1045ce558aff0083fdcbb69665ffbb30083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

**Comment by promag** [comment]
@MarcoFalke Apple LLVM version 10.0.1 (clang-1001.0.46.4) with `--enable-debug --with-sanitizers=thread`.

File: src/bench/rollingbloom.cpp [hunk 1]
```diff
@@ -28,4 +28,13 @@ static void RollingBloom(benchmark::State& state)
...
+static void RollingBloomReset(benchmark::State& state)
+{
+    CRollingBloomFilter filter(120000, 0.000001);
+    while (state.KeepRunning()) {
+        filter.reset();
+    }
+}
+
 BENCHMARK(RollingBloom, 1500 * 1000);
+BENCHMARK(RollingBloomReset, 150);
```
Comment (MarcoFalke) [inline src/bench/rollingbloom.cpp]: Please adjust this, so that it takes as much time as the other benchmarks (~4seconds, depending on arch)

**Comment by MarcoFalke** [comment]
Please mention that in the OP. Performance improvements when run with a sanitizer are completely meaningless to non-developers.

**Comment by laanwj** [comment]
Also, always benchmark in release mode. `--enable-debug` dials down the optimization level. It's very possible that you'll see no difference without that.

**Comment by promag** [comment]
After all in release mode I have the same result as @MarcoFalke. Updated OP.

File: src/bench/rollingbloom.cpp (hunk 1 above)
Comment (promag) [reply src/bench/rollingbloom.cpp]: Done.

**Comment by MarcoFalke** [comment]
re-utACK df9e15f092

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-utACK df9e15f092
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUipzwwAha/6gluoBoDmrtyN06+eiKHYnCrPhkDCFXmUptsHMMkCLnxpzWslG9zX
CXO6RZq48KQj6ZSx0TjdVFZuq5JEICqbiSk3u9McQxl2sILJx7x0fdz7yS/elu5m
RvT+cFIxav/eb7Bgnm6Zoz9EdRHFqyvglN3xBKowF5M2AHXbDq9trhM8oL9sgSFZ
dtHpaAern/L6s1UWF5zgVpDIqVB6Xr0mp0HodjMo8UurMTzB4nzZR6PAABJOeXPG
CGA7Xfkqj/gC/MP4gUKndS1aRlsY0uERPWclMd4NDF6fEX1mSAQHzJ7GO8pDBM5D
RwDpmV99ZdYUiD0Uv0IM2LfQQjWMSXNgoYm4CNf3R9FTLMUmM15xU8BcXTAXwLh2
0baT6VIdDhy12BtpcDm9cxrIoKtU5OBFxjBLRY7ZkyMs8LlEjBxadlSBBOlDEUsm
6uAj8b2Mek26y5fRJ+eBFok9B/up9G7ci5GuCmMhVf4syA1bSr+q/sIBr2U0Phea
TPfT5e04
=joN1
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `156b97ccbe8a8aa40230d38498141d8b529ac9c16533d56aa09c7bd632f6c12c  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108156b97ccbe8a8aa40230d38498141d8b529ac9c16533d56aa09c7bd632f6c12cf010a431293d45034a1ec70f8f71c4f9b54a08fff0102e03dfc42c2e7fe30451a3fc4bf258e308f1045ce56941f0085e792bc56e4300a70083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010337e6ec0685f87e9573c3bb5ae844aa608f1045ce56940f0085b2d74e10a2beb260083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010a047967f52bba35c3471d283e4f016c808f1045ce56941f008200c5da48aaf2f040083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ded12389cd0b7e6b21de9f420739688108f0109ffdac775699a5f34c29fafd31e228d808f1045ce56941f0087a5327d147afa5240083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

**Comment by jamesob** [comment]
re-utACK https://github.com/bitcoin/bitcoin/pull/16073/commits/df9e15f092c18a8047f09307576c2b77b9c8d01c

**Comment by practicalswift** [comment]
utACK d2dbc7da26e1ca40200521c05a0b1ca75578acd2

**Comment by sipa** [comment]
utACK d2dbc7da26e1ca40200521c05a0b1ca75578acd2

**Comment by MarcoFalke** [comment]
@sipa, @practicalswift Just FYI, you reviewed an incorrect commit and your ACKs haven't been recorded in the merge.

**Comment by promag** [comment]
Sort by date FTW.

**Comment by practicalswift** [comment]
@MarcoFalke Oh sorry about that and thanks for the notification!

**Pull request merged**